### PR TITLE
Make Benchmark Output Analysis Function Configurable

### PR DESCRIPTION
**Pull request by shawntabrizi**
This PR introduces a new CLI flag to the benchmarking pipeline:

```
	/// Which analysis function to use when outputting benchmarks:
	/// * min-squares (default)
	/// * median-slopes
	/// * max (max of min squares and median slopes for each value)
	#[structopt(long)]
	pub output_analysis: Option<String>,
```

This will allow a user to specify which analysis function to use when outputting benchmarking results, and even use the maximum value between the analysis functions.

For now only `min-squares` and `median-slopes` is supported, but this is extensible to support future analysis functions too.

### CODE REVIEW

File: frame/system/src/weights.rs [hunk 1]
```diff
@@ -55,79 +55,79 @@ pub trait WeightInfo {
 
 /// Weights for frame_system using the Substrate node and recommended hardware.
 pub struct SubstrateWeight<T>(PhantomData<T>);
-impl<T: crate::Config> WeightInfo for SubstrateWeight<T> {
+impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
```
Comment (shawntabrizi) [inline frame/system/src/weights.rs:58]:
```suggestion
impl<T: crate::Config> WeightInfo for SubstrateWeight<T> {
```

**Comment by shawntabrizi** [comment]
/benchmark runtime custom --chain=dev --steps=50 --repeat=20 --pallet=frame_system --extrinsic="*" --execution=wasm --wasm-execution=compiled --heap-pages=4096 --output=./frame/system/src/weights.rs --template=./.maintain/frame-weight-template.hbs --output-analysis=max

File: frame/system/src/weights.rs [hunk 2]
```diff
@@ -55,79 +56,79 @@ pub trait WeightInfo {
...
```
Comment (shawntabrizi) [inline frame/system/src/weights.rs:59]:
```suggestion
impl<T: crate::Config> WeightInfo for SubstrateWeight<T> {
```

File: utils/frame/benchmarking-cli/src/writer.rs [hunk 3]
```diff
@@ -150,11 +155,20 @@ fn extract_errors(model: &Option<RegressionModel>) -> impl Iterator<Item=u128> +
...
-	let extrinsic_time = Analysis::min_squares_iqr(&batch.results, BenchmarkSelector::ExtrinsicTime).unwrap();
-	let reads = Analysis::min_squares_iqr(&batch.results, BenchmarkSelector::Reads).unwrap();
-	let writes = Analysis::min_squares_iqr(&batch.results, BenchmarkSelector::Writes).unwrap();
+	let analysis_function = match analysis_choice {
+		AnalysisChoice::MinSquares => Analysis::min_squares_iqr,
+		AnalysisChoice::MedianSlopes => Analysis::median_slopes,
+		AnalysisChoice::Max => Analysis::max,
+	};
+
+	let extrinsic_time = analysis_function(&batch.results, BenchmarkSelector::ExtrinsicTime).unwrap();
```
Comment (thiolliere) [inline utils/frame/benchmarking-cli/src/writer.rs:169]: if linear regression fails this unwrap will be triggered.
Also having an expect with the reason would be better I think.

**Review by thiolliere — APPROVED** [review]
looks good to me,

But I think the code should make use of expect and Result instead of unwraps with no message.
Also the order of variables when doing the max of both analysis must be asserted IMO

File: frame/benchmarking/src/analysis.rs [hunk 4]
```diff
@@ -215,6 +248,34 @@ impl Analysis {
...
+	pub fn max(r: &Vec<BenchmarkResults>, selector: BenchmarkSelector) -> Option<Self> {
+		let median_slopes = Self::median_slopes(r, selector);
+		let min_squares = Self::min_squares_iqr(r, selector);
+
+		if median_slopes.is_none() || min_squares.is_none() {
+			return None;
+		}
+
+		let median_slopes = median_slopes.unwrap();
+		let min_squares = min_squares.unwrap();
```
[the remark this answers is not in the document]
Comment (shawntabrizi) [reply frame/benchmarking/src/analysis.rs:264]: I check these are not NONE above right? So unwrap here should be impossible to fail?

[the remark this answers is not in the document]
Comment (thiolliere) [reply frame/benchmarking/src/analysis.rs:264]: ah my bad I thought it was an `&&`

File: frame/benchmarking/src/analysis.rs [hunk 5]
```diff
@@ -31,13 +32,48 @@ pub struct Analysis {
...
+	MinSquares,
+	/// Use median slopes for analyzing the benchmarking results.
+	MedianSlopes,
+	/// Use the maximum values among all other analysis functions for the benchmarking results.
+	Max,
+}
+
+impl Default for AnalysisChoice {
+	fn default() -> Self {
+		AnalysisChoice::MinSquares
```
Comment (kianenigma) [inline frame/benchmarking/src/analysis.rs:55]: why isn't max the default, if we are to be pessimistic?

**Review by kianenigma — APPROVED** [review]

File: frame/benchmarking/src/analysis.rs (hunk 5 above)
Comment (shawntabrizi) [reply frame/benchmarking/src/analysis.rs:55]: its not clear to me that we should use `max` for everything. In general, min squares is a solid choice for an algorithm, this PR just gives us more options when there are weird results like we say with `remark_with_event`

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**